### ISSUE TITLE
Completion_course_user e-mails are not send out

### ISSUE THREAD

**Issue by michalludynia** [iomad/iomad#1278]
Completion_course_user emails seems not to work on IOMAD 3.7 stable. 

I do not see any course completion email's in outgoing email's report and they are not delivered to students. I can see completed status with date in course report, so messages should have been sent out.

- completion_course_user is enabled in emails templates
- cron is working
- warning emails (not started, not completed) related to course are sent out properly

Any help is much appreciated!
Thanks

**Comment by turf212** [comment]
Are they in a different language?  Do they appear in the outgoing email report?

**Comment by michalludynia** [comment]
All emails are in the same language and they are activated in emails settings for the company. 
Completion emails do not appear in outgoing   e-mail report. 

As I mentioned above the other emails like course not started or course completion warning works fine and I can see them in the report. 

Course is Licensed with instant access = false

**Comment by michalludynia** [comment]
Hi! @turf212 
I would like to ask if maybe you had reviewed that case and have some information about it?
Thanks!

**Comment by turf212** [comment]
There was a bug fixed and pushed yesterday under #1292 which was causing issues for some of the automated emails from being processed.  Can you please patch to the latest version and let me know if this has solved your issue?

**Comment by michalludynia** [comment]
@turf212 I've patched latest version today and the problem still occurs. In "course completion report" I can see status 100% and date completed, but user completion emails do not appear in outgoing email report and they are not send out physically.

**Comment by turf212** [comment]
Can you please check the following in your scheduled tasks (site admin -> server -> scheduled tasks):

- Email reports scheduled task is running and doesn't have a fail delay on it. (It should run once a day)
- If it is running change ensure you have PHP in your system paths (site admin -> server -> system paths) and manually run the email reports task (there will be a run now link)  Does it fail?

After this we are to to debugging the cron task in /local/email_reports/lib.php

**Comment by michalludynia** [comment]
@turf22 
I've check scheduled tasks and the email reports scheduled task is running properly. I've also manually run it and the course completion emails haven't been sent out. I have checked also the PHP path and there's correct path in there.

As I mentioned above "warning" and "not started" emails work properly. Only user course completion email seems not to work.

Really appreciate your help.

**Comment by turf212** [comment]
Ok.  To change tack on this, are the users being entered into the Iomad reports as being completed for the course or are they still in-progress there too.

If they are still in progress check that the course has completion criteria set up (Course admin -> course completion).  If there is nothing set as to why the course is completed the event doesn't get triggered so the emails are not sent and the users is not updated in the reports.

**Comment by michalludynia** [comment]
@turf212 
I can see progress bar on 100% status and I can also see completion date - screen attached. 
Course critera are defined properly. 

How it looks like:
User is enroled on the course by the license with "instant access" parameter set to false.
I have only one site inside course which is only one completion criterium defined. When the user click it, the status of completion change to "completed" and I can see completion date in report, but the user does not receive an email.

<img width="657" alt="Screenshot 2020-03-25 at 12 30 03" src="https://user-images.githubusercontent.com/61524640/77532161-9c04d900-6e94-11ea-8382-7b4529ed1450.png">

There is also no something like "sending completion course emails" while running scheduled task manually:

<img width="953" alt="Screenshot 2020-03-24 at 15 51 06" src="https://user-images.githubusercontent.com/61524640/77532583-70ceb980-6e95-11ea-8678-ee3bb036d9ce.png">

**Comment by turf212** [comment]
Sorry, that was my mistake.  The email gets sent out via a trapped event from local/iomad/lib/company.php public static function course_completed(\core\event\course_completed $event)

The only checking which is in that function is to see if there is already an entry in the mdl_local_iomad_track table which has the same enrolment timestamp and a real timecompleted timestamp so that resetting the course completion settings within the course don't cause the emails to be sent again for that user/manager.  This may be impacting on your testing if you didn't have the emails set up initially and the user was already marked complete on the course but that doesn't look to be the case from the image you have attached.

I'm running some tests on my 3.7 test site I'll let you know what I find.

**Comment by turf212** [comment]
Simple course with 1 user and course completion criteria set to be completion of a choice only.  Choice is completed when user makes one.

User chooses  choice option and is marked as complete on cron run
User is marked as course complete on next cron run
This triggers user to have completion report updated for this course in the Iomad reports
This then triggers email to be added to the Iomad email queue and email sent for processing

Everything looks fine.   This site is a two language install (English and Pirate) the user's default language is English, for which there are no changes to the templates.

If you can run this SQL against your DB it may help find out what is not set up right on your site:

select id,name,lang,companyid,disabled from mdl_email_template where name='completion_course_user';

Im assuming your table prefix is mdl_

**Comment by michalludynia** [comment]
<img width="1032" alt="Screenshot 2020-03-25 at 14 56 55" src="https://user-images.githubusercontent.com/61524640/77544171-f871f380-6ea8-11ea-9e80-6275cebd19e6.png">
That's the result of above query. 

Problem seems to occur only on licensed courses. When I changed course to self enrolment and "no sharing" the completion email was sent out.  

I'll come back with more info.

**Comment by michalludynia** [comment]
Got some new tests:

User completion e-mail works properly ONLY with "no licensed " courses. 
When I changed course settings in IOMAD COURSE SETTINGS from "self enrolment" to "licensed = yes" and no matter how the sharing option is set, the course completion email stops to be sent out. 

@turf212 Have you tested that case on yours test site?

**Comment by turf212** [comment]
There is some checking against license courses, specifically if the license allocated to the user is a program of courses.  If it is, and the user if completing the last license course in the program, then the system will send the user_programcompleted template instead.

I'll check the simple scenario you've outlined in any case.

**Comment by turf212** [comment]
Changed the course to a license based course and created a standard license and assigned it to a user.  Completed the course as that user and email sent our successfully.

**Comment by michalludynia** [comment]
I have tried almost everything and still completion course emails are not sent out on my instance no matter how I setup the license, email templates:

- I've created completely new company, course, license and user
- Scheduled task works every 3 minutes without delays.
- Tried multiple license configurations
- Tried default template of emails in english language

But nothing helps. 

Only when I change the course to "no licensed" the emails start to work perfeclty. The license causes a problem. 

 @turf212 Are there any steps I can do more? I'm thinking about installing completely new instance.

**Comment by turf212** [comment]
If you enable the user_programcompleted email does it send this out instead?

**Comment by turf212** [comment]
This is the lines of code where it may be falling down

https://github.com/iomad/iomad/blob/5e4126825278f40959340d22799b3a64a58c7647/local/iomad/lib/company.php#L3617-L3646

If you turn off the cron task so you can run it by hand and put some debug in around this bit of the function it may help pinpoint what scenario you are hitting.

**Comment by michalludynia** [comment]
> If you enable the user_programcompleted email does it send this out instead?

I've changed license type to "program of courses" = "yes" and user_programcompleted email is not send out as well :(

Thanks for code! I'll try to check it out.

**Comment by dsquare58** [comment]
![Screenshot (229)](https://user-images.githubusercontent.com/44236940/77974471-2adb8080-72ef-11ea-9d84-f618d208ef81.png)
Now This is what I get when I run crong.php

**Comment by michalludynia** [comment]
@turf212 
I do not know the moodle code well, so I fell a little bit lost in debugging this, but I can see that users appears correctly in mdl_course_completions table. 

I'm just all the time wondering why only with license the problem occurs. 
As I mentioned it does not matter if it's course program licensed or just single one, just license itself.
Maybe you can post your test configuration of license and company that worked in your test platform? 

Does Email reports task can be executed every 5 minutes instead once a day? That's my setup. 

Is there something more I can do, to verify, why only licensed courses has that problem? 

Really appreciate your help.

**Comment by michalludynia** [comment]
I've figured out which part of code stops completion email to be sent out. 

When I've commented that fragment of code lines 3579-3589 in 3.7 version (there are a little bit different lines in recent release of IOMAD) the completion emails started to work. 

@turf212 What can cause that case? 

https://github.com/iomad/iomad/blob/5e4126825278f40959340d22799b3a64a58c7647/local/iomad/lib/company.php#L3579-L3589
